File: scripts/news_bot.py
import feedparser
import json
import os
from bs4 import BeautifulSoup
from datetime import datetime
from time import mktime

# Lista de feeds RSS por categoria
FEEDS = {
    "Agro": "https://g1.globo.com/rss/g1/economia/agronegocios/",
    "Economia": "https://g1.globo.com/rss/g1/economia/",
    "Negócios": "https://www.infomoney.com.br/negocios/feed/",
    "Mercados": "https://br.investing.com/rss/market_overview.rss"
}

def extract_image(entry):
    # Tenta media_content
    if 'media_content' in entry and len(entry.media_content) > 0:
        return entry.media_content[0]['url']
    
    # Tenta enclosure
    if 'enclosures' in entry and len(entry.enclosures) > 0:
        for enc in entry.enclosures:
            if enc.type and 'image' in enc.type or enc.href.endswith(('.jpg', '.png', '.jpeg')):
                return enc.href

    # Tenta buscar na description ou content
    html_content = ""
    if 'content' in entry:
        html_content = entry.content[0].value
    elif 'description' in entry:
        html_content = entry.description
        
    if html_content:
        soup = BeautifulSoup(html_content, 'html.parser')
        img_tag = soup.find('img')
        if img_tag and img_tag.get('src'):
            return img_tag['src']
            
    return "https://images.unsplash.com/photo-1590283603385-17ffb3a7f29f?ixlib=rb-4.0.3&auto=format&fit=crop&w=800&q=80" # Fallback corporativo

def extract_summary(entry):
    html_content = entry.get('description', '')
    soup = BeautifulSoup(html_content, 'html.parser')
    text = soup.get_text(separator=' ', strip=True)
    if len(text) > 180:
        return text[:177] + "..."
    return text

def parse_feeds():
    all_news = []
    for category, url in FEEDS.items():
        print(f"Buscando feed de {category}: {url}")
        feed = feedparser.parse(url)
        for entry in feed.entries[:8]: # Pega as 8 últimas de cada categoria
            try:
                published_parsed = entry.get('published_parsed')
                if published_parsed:
                    dt = datetime.fromtimestamp(mktime(published_parsed))
                    date_str = dt.strftime("%d/%m/%Y")
                else:
                    date_str = datetime.now().strftime("%d/%m/%Y")
                    
                image = extract_image(entry)
                summary = extract_summary(entry)
                
                if not summary or not entry.get('title'):
                    continue
                    
                news_item = {
                    "id": entry.get('id', entry.link),
                    "title": entry.title,
                    "link": entry.link,
                    "summary": summary,
                    "image": image,
                    "category": category,
                    "date": date_str
                }
                all_news.append(news_item)
            except Exception as e:
                print(f"Erro ao processar notícia: {e}")
                
    return all_news

def main():
    print("Iniciando Bot de Notícias...")
    news = parse_feeds()
    
    # Sort news to mix categories nicely (or just keep them appended)
    # We will just reverse it so the latest from the loops don't dominate the top
    # Or just keep it as is, frontend can sort
    
    output_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "src", "data")
    os.makedirs(output_dir, exist_ok=True)
    
    output_file = os.path.join(output_dir, "news.json")
    
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(news, f, ensure_ascii=False, indent=2)
        
    print(f"Sucesso! {len(news)} notícias salvas em {output_file}")

if __name__ == "__main__":
    main()

File: scripts/news_bot.mjs
import fs from 'fs';
import path from 'path';
import Parser from 'rss-parser';

const parser = new Parser({
  customFields: {
    item: ['media:content', 'description', 'content:encoded']
  },
  requestOptions: {
    headers: { 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64 AppleWebKit/537.36)' }
  }
});

const FEEDS = [
  { category: 'Economia', url: 'https://g1.globo.com/rss/g1/economia/' },
  { category: 'Agro', url: 'https://g1.globo.com/rss/g1/economia/agronegocios/' },
  { category: 'Agro', url: 'https://www.canalrural.com.br/feed/' }
];

function extractImage(item) {
  if (item.enclosure && item.enclosure.url && item.enclosure.type && item.enclosure.type.includes('image')) {
    return item.enclosure.url;
  }
  if (item['media:content'] && item['media:content'].$) {
    return item['media:content'].$.url;
  }
  const content = item['content:encoded'] || item.content || item.description || '';
  const imgMatch = /<img.*?src=["'](.*?)["']/.exec(content);
  if (imgMatch) {
    return imgMatch[1];
  }
  return null; // NO FALLBACKS!
}

function cleanSummary(text) {
  if (!text) return '';
  let clean = text.replace(/<[^>]*>?/gm, '').trim();
  clean = clean.replace(/&#8211;/g, '-').replace(/&nbsp;/g, ' ');
  if (clean.length > 160) {
    return clean.substring(0, 157) + '...';
  }
  return clean;
}

async function main() {
  console.log('Iniciando Robô de Notícias em Node.js (Sem Imagens Genéricas)...');
  let allNews = [];
  
  for (const feed of FEEDS) {
    console.log(`Buscando feed de ${feed.category} em: ${feed.url}`);
    try {
      // Manual fetch and trim to bypass G1 XML spacing issues
      const response = await fetch(feed.url, {
        headers: { 'User-Agent': 'Mozilla/5.0' }
      });
      let xml = await response.text();
      xml = xml.trim();
      const parsedFeed = await parser.parseString(xml);
      
      let count = 0;
      for (const item of parsedFeed.items) {
        if (count >= 5) break; // Limit per feed
        
        const image = extractImage(item);
        // CRITICAL: Only accept news that actually have a REAL image from the agency
        if (!image) continue;

        const summary = cleanSummary(item.contentSnippet || item.content || item.description);
        
        let dateStr = new Date().toLocaleDateString('pt-BR');
        if (item.pubDate) {
          const dt = new Date(item.pubDate);
          if (!isNaN(dt.getTime())) dateStr = dt.toLocaleDateString('pt-BR');
        }

        if (item.title && !allNews.some(n => n.title === item.title)) {
          allNews.push({
            id: item.link || Math.random().toString(),
            title: item.title,
            link: item.link || '#',
            image,
            summary: summary || 'Clique para ler a matéria completa...',
            category: feed.category,
            date: dateStr
          });
          count++;
        }
      }
    } catch (e) {
      console.error(`Erro ao buscar ${feed.category}:`, e.message);
    }
  }
  
  const mixedNews = [];
  const maxItems = Math.max(...FEEDS.map(() => 5));
  for (let i = 0; i < maxItems; i++) {
    for (let j = 0; j < FEEDS.length; j++) {
      const catNews = allNews.filter(n => n.category === FEEDS[j].category);
      if (catNews[i] && !mixedNews.includes(catNews[i])) mixedNews.push(catNews[i]);
    }
  }
  
  const outputDir = path.join(process.cwd(), 'src', 'data');
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }
  
  const outputPath = path.join(outputDir, 'news.json');
  fs.writeFileSync(outputPath, JSON.stringify(mixedNews, null, 2), 'utf-8');
  console.log(`Sucesso! ${mixedNews.length} notícias 100% REAIS atualizadas em ${outputPath}.`);
}

main();
